### PR TITLE
StatelessSession.fetch()

### PR DESCRIPTION
**Pull request by gavinking**
Draft of a `fetch()` operation for `StatelessSession` (https://hibernate.atlassian.net/browse/HHH-15001)

### CODE REVIEW

**Review by sebersole — APPROVED** [review]
LGTM

**Comment by sebersole** [comment]
Just pushed a fix to a breakage I caused in the build.  Sorry.  Can you rebase this from upstream wip/6.0 and push again?

**Comment by gavinking** [comment]
Does anyone have any suggestions about how this should be tested? I would prefer to add tests to something that already exists? But what, precisely?

**Comment by sebersole** [comment]
> how this should be tested

I'm guessing you mean something more than "in a JUnit test".  But not sure what you are asking

> add tests to something that already exists

What kind of "something"?  You mean an existing test class?

**Comment by gavinking** [comment]
Yes, an existing class.

**Comment by sebersole** [comment]
It's a completely new method/feature.  Why would you not want a new test class?

**Comment by gavinking** [comment]
Ooops, I had forgotten about this one .... but I still want it!

File: hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java [hunk 1]
```diff
@@ -358,6 +360,56 @@ private Object createProxy(EntityKey entityKey) {
...
+			}
+		}
+		else if ( association instanceof PersistentCollection) {
+			PersistentCollection<?> persistentCollection = (PersistentCollection<?>) association;
+			if ( !persistentCollection.wasInitialized() ) {
+				CollectionPersister persister =
+						getFactory().getMetamodel().collectionPersister( persistentCollection.getRole() );
+				Object key = persistentCollection.getKey();
+				persistenceContext.addUninitializedCollection( persister, persistentCollection, key );
+				persistentCollection.setCurrentSession(this);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java:400]: Shouldn't we also unset the session here in the finally block?

**Review by beikov — COMMENTED** [review]
Looks good except for the collection initialization case, but the formatting could be improved ;)

**Comment by gavinking** [comment]
Don't merge this yet. Something changed in H6 and it broke this PR.

I will fix it when I have time again.

File: hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java (hunk 1 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java:400]: Not as far as I am aware: I don't think SS generally cleans up it's temporary PC in a finally. (Remember: failures are unrecoverable in Hibernate, and the session is considered hosed after a failure.)

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java:400]: Dunno, it just felt odd that you unset it in the above if-branch but not here. If you say it's fine, I'm ok with that. Just wanted to bring your attention to this.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/internal/StatelessSessionImpl.java:400]: Oh you're saying I forget to unset it, not that you want me to unset it in a `finally`. Yeah, sure, I guess that would be more consistent.

**Comment by gavinking** [comment]
See [HHH-15001](https://hibernate.atlassian.net/browse/HHH-15001).

**Comment by gavinking** [comment]
Alright, at last this is fixed and ready!

**Pull request merged**